src/Microsoft.Data.Analysis/DataFrameRowCollection.cs

```diff
@@ -0,0 +1,130 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.Data.Analysis
+{
+
+    /// <summary>
+    /// Represents the rows of a <see cref="DataFrame"/>
+    /// </summary>
+    public class DataFrameRowCollection : IEnumerable<DataFrameRow>
+    {
+        private long _currentIndex = 0;
+        private DataFrame _dataFrame;
+
+        /// <summary>
+        /// Initializes a <see cref="DataFrameRowCollection"/>.
+        /// </summary>
+        internal DataFrameRowCollection(DataFrame dataFrame)
+        {
+            _dataFrame = dataFrame;
+        }
+
+        /// <summary>
+        /// An indexer to return the <see cref="DataFrameRow"/> at <paramref name="index"/>
+        /// </summary>
+        /// <param name="index">The row index</param>
+        public DataFrameRow this[long index]
+        {
+            get
+            {
+                List<object> row = GetRow(index);
+                return new DataFrameRow(row);
+            }
+        }
+
+        internal List<object> GetRow(long rowIndex)
+        {
+            var ret = new List<object>();
+            var columns = _dataFrame.Columns;
+            foreach (var column in columns)
+            {
+                ret.Add(column[rowIndex]);
+            }
+            return ret;
+        }
+
+        /// <summary>
+        /// Returns an enumerator of <see cref="DataFrameRow"/> objects
+        /// </summary>
+        public IEnumerator<DataFrameRow> GetEnumerator()
+        {
+            DataFrameRow _currentRow = new DataFrameRow(null);
+            while (_currentIndex < _dataFrame.RowCount)
+            {
+                _currentRow._row = GetRow(_currentIndex);
+                yield return _currentRow;
+                _currentIndex++;
+            }
+        }
+
+        /// <summary>
+        /// The number of rows in this <see cref="DataFrame"/>.
+        /// </summary>
+        public long Length => _dataFrame.RowCount;
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+    }
+
+
+    /// <summary>
+    /// A DataFrameRow is a collection of values that represent a row in a <see cref="DataFrame"/>.
+    /// </summary>
+    public class DataFrameRow : IEnumerable<object>
+    {
+        internal List<object> _row;
```

Comment:
Instead of copying all the values out into this object, I wonder if it would be better if `DataFrameRow` was more like a "view" into the current values of the `DataFrame`. That way it could also be editable.

To do this, `DataFrameRow` would get a reference to the `DataFrame` and a `rowIndex`. Any time someone asked for a value or to set a value, it would just forward that to the correct underlying value inside the actual `DataFrame`. Thoughts?

Reply:
I like this idea 👍 In this case _View_ should get part of the name, to make this explicit -> `DataFrameRowView`.

Just thinking about use cases where one really wants to have a copy, so that modifications don't "flow" back to the actual dataframe. Here I have some heuristics in mind, that manipulate the data, get the score and if it's worse, switch back to previous values.

So maybe it will be good to provide both? I.e. like here with the copy, and one with the view?

Reply:
> In this case View should get part of the name, to make this explicit -> DataFrameRowView.

I am not sure. We don't have the word View in `DataFrameColumn`. My thinking is that `DataFrameRow` and `DataFrameColumn` would work similarly.

> So maybe it will be good to provide both?

For the copy case - couldn't the user copy the data out themselves?

Reply:
> `DataFrameRow` and `DataFrameColumn` would work similarly.

That's fine and the least astonishing.

> couldn't the user copy the data out themselves?

You're right. That will be pretty trivial.

Reply:
I agree with the idea here. Making `DataFrameRow` a view into the `DataFrame` sounds like a good idea. Updated the PR.